### ISSUE TITLE
support modules as parameters in generic functions

### ISSUE THREAD

**Issue by ZeeD** [python/mypy#6132]
Note: if you are reporting a wrong signature of a function or a class in
the standard library, then the typeshed tracker is better suited
for this report: https://github.com/python/typeshed/issues

Please provide more information to help us understand the issue:

* Are you reporting a bug, or opening a feature request?

I think this is a bug (the behaviour is consistent with other kind of objects)

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

```
import typing
T = typing.TypeVar('T')
def same(obj: T) -> T:
    return obj
add = same(3).__add__
TypeVar = same(typing).TypeVar
```

* What is the actual behavior/output?

```
minimal.py:6: error: Module has no attribute "TypeVar"
```

* What is the behavior/output you expect?

[no errors] 

* What are the versions of mypy and Python you are using?
  Do you see the same issue after installing mypy from Git master?

0.650

* What are the mypy flags you are using? (For example --strict-optional)

```
[mypy]
disallow_untyped_defs = True
```

**Comment by ZeeD** [comment]
I understand that the type of all the modules is "Module", but it whould be useful to "propagate" the specific names exported by the specific type

**Comment by ilevkivskyi** [comment]
> I understand that the type of all the modules is "Module", but it whould be useful to "propagate" the specific names exported by the specific type

Yes. This would be a useful feature. It was discussed a while ago in the context of protocols.

**Comment by ilevkivskyi** [comment]
Although the original example doesn't give an error anymore for me, it still doesn't work properly.

**Comment by ZeeD** [comment]
yep, the behavior seems slightly changed ... seems that the keys are "found" but are marked as "Any"; a new example is here ([mypy play](https://mypy-play.net/?mypy=latest&python=3.12&gist=1039dda923e84715f52b6788731a9f8c))

```python
import typing

def same[T](obj: T) -> T:
    return obj

int_orig = 3
int_copy = same(int_orig)
module_orig = typing
module_copy = same(module_orig)

typing.reveal_type(int_orig)
typing.reveal_type(int_copy)
typing.reveal_type(int_orig.__add__)
typing.reveal_type(int_copy.__add__)

typing.reveal_type(module_orig)
typing.reveal_type(module_copy)
typing.reveal_type(module_orig.TypeVar)
typing.reveal_type(module_copy.TypeVar)
```

with this output:

```
main.py:11: note: Revealed type is "builtins.int"
main.py:12: note: Revealed type is "builtins.int"
main.py:13: note: Revealed type is "def (builtins.int) -> builtins.int"
main.py:14: note: Revealed type is "def (builtins.int) -> builtins.int"
main.py:16: note: Revealed type is "types.ModuleType"
main.py:17: note: Revealed type is "types.ModuleType"
main.py:18: note: Revealed type is "def (name: builtins.str, *constraints: Any, bound: Union[Any, None] =, covariant: builtins.bool =, contravariant: builtins.bool =, infer_variance: builtins.bool =) -> typing.TypeVar"
main.py:19: note: Revealed type is "Any"
Success: no issues found in 1 source file
```
